### ISSUE TITLE
[elastic_agent/fleet_server] Discuss: Align versioning with the stack

### ISSUE THREAD

**Issue by ruflin** [elastic/integrations#2147]
Some packages like apm have been aligned with the stack version. I expect updates to elastic_agent and fleet_server to not happen very often but would like to see them aligned with the stack version. This should make it easier to understand which version is compatible with which stack version. 

Proposal:

* Release a package 7.16.0 which is compatible with `^7.16.0 || ^8.0.0`
* Release a package 8.0.0 which is compatible with `^8.0.0`

No further releases until we need to change something. It is likely that 7.16.0 package will not get any updates. 

In theory, the 8.0.0 package is not needed and it would make sure we can still easily release changes to 7.16.0 without having to branch integrations repo or similar. But from a user perspective it feels more natural to have the 8.0 package around. 

There is another option to make the branching easier. The fleet_server package could be moved to the fleet_server repository and released from there. Same for elastic_agent package which could be moved into the source of elastic_agent.

**Comment by ruflin** [comment]
@andresrc @jsoriano @mtojek Would be great to get your opinion on the above.

**Comment by mtojek** [comment]
and @jlind23

**Comment by jlind23** [comment]
@ruflin as said recently by @mtojek and @jsoriano, if those specific packages are bound to the stack, do we really need to serve them as packages or should we bundle them inside the stack?

**Comment by mtojek** [comment]
I'm concerned that this proposal violates all our principles in terms of CI/CD and free package publishing rules.

If this is REALLY the way we want to proceed we need to say it loud that packages will be also part of our release trains. I'm saying this because it would look like we're burning cycles to implement an automatic pipeline, but in reality, we just prefer releasing boxes.

**Comment by ruflin** [comment]
I don't think it is either or. All the packages should be in the registry, also the one bundled. This has some advantages. For example if there is a bug in 8.1.0 package of apm, 8.1.1 package can be released and the package is upgraded before having to upgrade Elasticsearch too. Also I like to see that we make sure these packages go through the same validation process.

In 7.16 there are no bundled packages.

> If this is REALLY the way we want to proceed we need to say it loud that packages will be also part of our release trains. I'm saying this because it would look like we're burning cycles to implement an automatic pipeline, but in reality, we just prefer releasing boxes.

There is a small minority of packages that don't follow the usual pipeline. These are the ones we are discussing here. This does not have any effect on the overall pipeline.

> I'm concerned that this proposal violates all our principles in terms of CI/CD and free package publishing rules.

What is this rule? Can you reference it?

**Comment by mtojek** [comment]
> I don't think it is either or. All the packages should be in the registry, also the one bundled. This has some advantages. For example if there is a bug in 8.1.0 package of apm, 8.1.1 package can be released and the package is upgraded before having to upgrade Elasticsearch too. Also I like to see that we make sure these packages go through the same validation process.

Could you please share the proposal when we decided on introducing bundled packages? I might have missed it. I'm wondering if we discussed all options and exhausted the topic with a conclusion that the full CI/CD is a bad idea for them.

> There is a small minority of packages that don't follow the usual pipeline. These are the ones we are discussing here. This does not have any effect on the overall pipeline.

This is also a concern as other package owners will consider following this path to exploit breaking changes or compatibility issues.

> What is this rule? Can you reference it?

We agreed that the package version doesn't mean anything else, but just follow semver rules. For compatibility with stack we have kibana version constraints and probably a similar property for elastic-agent in the future. 

Do we really need to align constraints with package versions? Why are version constraints insufficient? What do we do if we if there are 30 bugs to fix per stack release? Usually, we just published a package update independently.

Personal note:

I'm strongly in favor of moving towards the full CI/CD and preventing the idea of box releases no matter what. I just consider the automatic approach as more convenient, flexible, easier for developers. I'm fine with passing the reviewer role to another person that used to operate in the release train mode.
